doc/adr/0016-state-machines-for-UI-state.md

```diff
@@ -0,0 +1,59 @@
+# 0016 State machines for UI state management
+
+Date: Unknown
+
+## Status
+
+Accepted
+
+## Context
+
+Modern (JavaScript-based) User Interfaces have many components and lots of
+concurrent actions going on.  Due to the combination and number of possible
+actions being triggered, the integration between components has become
+increasingly complex over the years -- especially when considering
+asynchronicity and error handling.
+
+The JavaScript/front-end development community has increasingly been using
+[state machines](https://en.wikipedia.org/wiki/Finite-state_machine) and
+[state charts](https://en.wikipedia.org/wiki/UML_state_machine) to solve this
+issue of complexity.
+
+[XState](https://stately.ai/docs/xstate) is the - by far - largest state
+machine engine in the JavaScript ecosystem, featuring over 1 million downloads
+per week (ref: 2023-12-02).  It implements the *state chart* state machine
+type.  XState however is an extensive library with an extensive 'DSL' to
+describe state machines.  There are many alternatives implementing state charts
+or state machines that are less extensive.  One such example is [robot3](https://thisrobot.life/)
+which is extremely small, supports sub-state machines and features integration
+for React, PReact and Lit.  Its DSL is similarly small and thereby easy to
+read.  None of the integrations are of interest for LedgerSMB, but they
+indicate interest from a wider community.
+
+Considering that with its move to Vue, LedgerSMB is going in the same
+direction as other projects -- featuring a heavy (JavaScript-based) client
+application -- the project similarly needs a way to control the expected
+complexity.
+
+## Decision
+
+The web-based client application will use state machines to make state
+management as well as state transition explicit.  Similar to UI composition
+with components, will different parts of the UI be governed by their own state
+machines.
+
+The state machine library used by the project will be robot3, which uses XState
+as an example, but chooses to be as lean as possible in its implementation.
+
+## Consequences
+
+1. Any view or component will come with a state machine to describe and manage
+   its internal state and the allowed transitions
+2. Components don't change their own state: in response to events (their own or
```

Comment:
I think of a component as a visual element or widget, something used to display some subset of the fields/information about something else, which I'm calling here a document.

Components would be:
- Address widget
- Line item widget
- Collection of line items
- Collection of order adjustments
- Order display (combining many of those)
- Action buttons (to trigger state transitions)
- Filters (to filter a list of documents)

... there is some internal/local state on many of these (read-only, disabled, active, selected, etc) but not to the level they are linked directly to a state machine -- their state comes from the state of the parent document.

"Documents" would be:

- Sales invoice
- Vendor invoice
- AR transaction
- AP transaction
- GL transaction
- Sales Order
- Purchase Order

... etc.

These would be the objects tightly linked to one (or more) state machines, with an active state.